compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -599,6 +601,26 @@ object Objects:
 
           case _ => a
 
+    def filterType(tpe: Type)(using Context): Value =
+      val baseClasses = tpe.baseClasses
+      if baseClasses.isEmpty then a
+      else tpe match
+        case t @ SAMType(_, _) if a.isInstanceOf[Fun] => a // if tpe is SAMType and a is Fun, allow it
+        case _ => filterClass(baseClasses.head) // could have called ClassSymbol, but it does not handle OrType and AndType
+
+    def filterClass(sym: Symbol)(using Context): Value =
+        if !sym.isClass then a
+        else
+          val klass = sym.asClass
+          a match
+            case Cold => Cold
+            case ref: Ref if ref.klass.isSubClass(klass) => ref
```

Comment:
For consistency with other cases, I'd move the `if` inside the case body and combine with the following case.